File: backend/app/services/fastclaw_service.py
import asyncio
import json
import logging
import re
import time
from dataclasses import dataclass
from typing import Any, AsyncGenerator, Dict, List, Optional, Tuple

import httpx

logger = logging.getLogger(__name__)

# FastClaw /api/chat/stream 请求超时（秒）。agent 思考/工具执行可能很久，
# 连接阶段给 15s，读阶段不设上限（由调用方按空闲超时自行中止）。
CONNECT_TIMEOUT = 15.0

# agent_id → FastClaw 真实名字的进程内 TTL 缓存（key: (base_url, api_key, agent_id)）。
# 用于给存量配置回填可读名字：成功缓存 5 分钟，失败（FastClaw 不可达）30 秒内不重试。
_AGENT_NAME_CACHE: Dict[Tuple[str, str, str], Tuple[Optional[str], float]] = {}
_AGENT_NAME_CACHE_TTL = 300.0
_AGENT_NAME_FAIL_TTL = 30.0


class FastClawAgentError(Exception):
    """FastClaw Agent 调用失败（由 SSE 端点捕获后以 error 事件传导到前端）。"""


@dataclass
class FastClawRuntimeConfig:
    """一次 Agent 调用所需的运行时配置（由 StageConfig 解析而来）。"""

    base_url: str = ""
    api_key: str = ""
    agent_id: str = ""
    # 上游 end-user 标识（FastClaw 据此隔离会话/记忆/用量）；建议传「应用名-本地用户ID」
    end_user: str = ""


_IMAGE_URL_RE = re.compile(r"!\[[^\]]*\]\((https?://[^)\s]+)\)|(https?://[^\s)\]]+\.(?:png|jpe?g|webp|gif)[^\s)\]]*)")


def extract_image_url(text: str) -> Optional[str]:
    """从 agent 最终文本中提取第一张图片 URL（支持 markdown 图片语法与裸 URL）。"""
    if not text:
        return None
    m = _IMAGE_URL_RE.search(text)
    if not m:
        return None
    return m.group(1) or m.group(2)


class FastClawAgentService:
    """FastClaw Agent 调用代理。

    调用 FastClaw 仪表盘 SSE 接口 `POST /api/chat/stream`（Bearer API Key 鉴权），
    将内部的富事件流（content_delta / tool_call / tool_result / status /
    subagent_progress / error / done）归一化为统一的 dict 事件流。
    """

    async def run_agent(
        self,
        config: FastClawRuntimeConfig,
        message: str,
        session_key: str,
        images: Optional[List[str]] = None,
        params: Optional[Dict[str, Any]] = None,
    ) -> AsyncGenerator[Dict[str, Any], None]:
        """调用 agent 并产出归一化事件流。

        每个事件为 dict：
        - {"type": "content_delta", "data": {"delta": str}}
        - {"type": "tool_call", "data": {"id", "name", "arguments"}}
        - {"type": "tool_result", "data": {"id", "name", "result"}}
        - {"type": "status", "data": {"message": str}}
        - {"type": "subagent_progress", "data": {...}}
        - {"type": "error", "data": {"message": str}}
        - {"type": "done", "data": {}}
        """
        if not config.base_url or not config.api_key or not config.agent_id:
            raise FastClawAgentError("FastClaw Agent 配置不完整（base_url / api_key / agent_id）")

        url = config.base_url.rstrip("/") + "/api/chat/stream"
        body: Dict[str, Any] = {
            "agentId": config.agent_id,
            "sessionId": session_key,
            "message": message,
        }
        if images:
            body["imageUrls"] = images
        if params:
            body["params"] = params

        headers = {
            "Authorization": f"Bearer {config.api_key}",
            "Content-Type": "application/json",
            "Accept": "text/event-stream",
        }
        if config.end_user:
            # 上游 end-user：让 FastClaw 将会话/记忆/用量按「应用用户」隔离
            headers["X-Fastclaw-End-User"] = config.end_user

        timeout = httpx.Timeout(connect=CONNECT_TIMEOUT, read=None, write=30.0, pool=30.0)
        try:
            async with httpx.AsyncClient(timeout=timeout, follow_redirects=True) as client:
                async with client.stream("POST", url, headers=headers, json=body) as resp:
                    if resp.status_code != 200:
                        detail = ""
                        try:
                            payload = await resp.aread()
                            detail = payload.decode("utf-8", errors="replace")[:300]
                        except Exception:
                            pass
                        raise FastClawAgentError(
                            f"FastClaw /api/chat/stream 返回 HTTP {resp.status_code}: {detail}"
                        )
                    saw_delta = False
                    async for event in _iter_sse(resp.aiter_lines()):
                        async for norm in _normalize_event(event):
                            # FastClaw 会先流式发 content_delta，再在本轮结束补发一条完整 content；
                            # 若本轮已收到增量，跳过该条完整 content，避免前端重复拼接文本。
                            # 工具调用/状态事件意味着新一轮助手回复开始（可能走非流式路径），
                            # 此时重置标记，让新一轮的完整 content 正常透传。
                            if norm["type"] == "content_delta":
                                saw_delta = True
                            elif norm["type"] in ("tool_call", "tool_result", "status", "subagent_progress"):
                                saw_delta = False
                            if norm["type"] == "content" and saw_delta:
                                continue
                            yield norm
        except httpx.HTTPError as exc:
            logger.warning("FastClaw Agent 调用失败: %s", exc)
            raise FastClawAgentError(f"FastClaw Agent 调用失败: {exc}") from exc

    async def list_agents(
        self,
        base_url: str,
        api_key: str,
        end_user: str = "",
        timeout: float = CONNECT_TIMEOUT,
    ) -> List[Dict[str, Any]]:
        """列出该 API Key 可访问的 FastClaw agent（供 admin 页面拉取候选）。

        优先调 FastClaw dashboard 接口 `GET /api/agents`——它返回 agent 的真实
        名字（AgentRecord.name）；上游 `GET /v1/agents` 的 name 字段与 id 相同
        （buildAgentList 硬编码 `"name": ag.Name()`），对人不可读。

        兼容性回退：/api/agents 不存在（旧版本）、报错或返回空列表时回退到
        /v1/agents（此时 name=id）。两者都**不能**带 X-Fastclaw-End-User 头，
        否则 FastClaw 会切到无 Agent 的 app-user 空间导致列表为空。
        """
        base = base_url.rstrip("/")
        headers = {"Authorization": f"Bearer {api_key}", "Accept": "application/json"}
        if end_user:
            headers["X-Fastclaw-End-User"] = end_user
        async with httpx.AsyncClient(timeout=timeout) as client:
            # 1) dashboard 接口：带真实名字。仅当返回 200 且列表非空才采用，
            #    空列表不信任（agent 类型 Key 的「所属账号」可能没有 agent，
            #    但 /v1/agents 仍能按 ACL 列出绑定的 agent）。
            try:
                resp = await client.get(base + "/api/agents", headers=headers)
                if resp.status_code == 200:
                    data = resp.json()
                    agents = data.get("agents") if isinstance(data, dict) else None
                    if agents:
                        return [
                            {
                                "id": a.get("id", "") if isinstance(a, dict) else "",
                                "name": (a.get("name") or a.get("id") or "") if isinstance(a, dict) else "",
                                # 该接口的 model 仅来自 agent 级配置（通常为空），
                                # 与 /v1/agents 的全量解析结果口径不同；前端暂不使用
                                "model": a.get("model", "") if isinstance(a, dict) else "",
                            }
                            for a in agents
                        ]
            except (httpx.HTTPError, ValueError):
                pass  # 网络/JSON 解析失败走回退分支
            # 2) 上游接口回退：name 与 id 相同
            try:
                resp = await client.get(base + "/v1/agents", headers=headers)
            except httpx.HTTPError as exc:
                raise FastClawAgentError(f"连接 FastClaw 失败: {exc}") from exc
            if resp.status_code != 200:
                raise FastClawAgentError(f"FastClaw /v1/agents 返回 HTTP {resp.status_code}")
            data = resp.json()
            return data.get("agents", []) if isinstance(data, dict) else []

    async def resolve_agent_name(
        self,
        base_url: str,
        api_key: str,
        agent_id: str,
        timeout: float = 3.0,
    ) -> Optional[str]:
        """解析 agent_id 对应的 FastClaw 真实名字（AgentRecord.name，如 "Xulei"）。

        带进程内 TTL 缓存（成功 5 分钟 / 失败 30 秒）；FastClaw 不可达或未找到时
        返回 None 且短时间内不重试，用于给存量配置回填可读名字（best-effort，不抛错）。
        """
        if not agent_id:
            return None
        key = (base_url.rstrip("/"), api_key, agent_id)
        now = time.monotonic()
        cached = _AGENT_NAME_CACHE.get(key)
        if cached and cached[1] > now:
            return cached[0]
        name: Optional[str] = None
        try:
            agents = await self.list_agents(
                base_url=base_url, api_key=api_key, timeout=timeout
            )
            for a in agents:
                if isinstance(a, dict) and a.get("id") == agent_id:
                    candidate = a.get("name")
                    # 仅接受与 id 不同的可读名字：/v1/agents 回退路径 name==id，
                    # 此时视为未解析（返回 None），避免把 agt_xxx 当名字持久化
                    if isinstance(candidate, str) and candidate and candidate != agent_id:
                        name = candidate
                    break
        except Exception:
            name = None
        _AGENT_NAME_CACHE[key] = (
            name,
            now + (_AGENT_NAME_CACHE_TTL if name else _AGENT_NAME_FAIL_TTL),
        )
        return name


async def _iter_sse(lines):
    """按 SSE 块累积解析 data: 行（忽略 id:/event:/注释行，支持多行 data）。"""
    data_lines: List[str] = []
    async for line in lines:
        line = line.rstrip("\r")
        if line == "":
            if data_lines:
                payload = "\n".join(data_lines)
                data_lines = []
                yield payload
            continue
        if line.startswith("data:"):
            data_lines.append(line[5:].lstrip(" "))
        # 其他行（id: / event: / : ping）忽略


async def _normalize_event(payload: str):
    """解析 FastClaw 事件 JSON 并归一化为内部事件 dict。"""
    try:
        evt = json.loads(payload)
    except (ValueError, TypeError):
        # 非 JSON 行（如代理注入的杂讯）直接忽略，不中断流
        return
    if not isinstance(evt, dict):
        return
    etype = evt.get("type", "")
    data = evt.get("data") or {}
    if not isinstance(data, dict):
        data = {}

    if etype == "content_delta":
        delta = data.get("delta") or ""
        if delta:
            yield {"type": "content_delta", "data": {"delta": delta}}
    elif etype in ("content", "message"):
        content = data.get("content") or data.get("text") or ""
        if content:
            # 与 content_delta 区分：标记为 content（完整文本），由调用方按需去重
            yield {"type": "content", "data": {"delta": content}}
    elif etype == "tool_call":
        yield {
            "type": "tool_call",
            "data": {
                "id": data.get("id", ""),
                "name": data.get("name", ""),
                "arguments": data.get("arguments", ""),
            },
        }
    elif etype == "tool_result":
        yield {
            "type": "tool_result",
            "data": {
                "id": data.get("id", ""),
                "name": data.get("name", ""),
                "result": data.get("result", ""),
            },
        }
    elif etype in ("tool_progress", "status"):
        yield {"type": "status", "data": {"message": data.get("message") or data.get("tool") or str(data)}}
    elif etype == "subagent_progress":
        yield {"type": "subagent_progress", "data": data}
    elif etype == "error":
        yield {"type": "error", "data": {"message": data.get("message") or str(data)}}
    elif etype == "done":
        yield {"type": "done", "data": {}}
    # 其余事件（turn_pending / steer 等）忽略


fastclaw_agent_service = FastClawAgentService()

File: backend-ts/src/services/fastclaw-service.ts
/**
 * FastClaw Agent 调用代理（原样移植 Python `app/services/fastclaw_service.py`）。
 *
 * 调用 FastClaw 仪表盘 SSE 接口 `POST /api/chat/stream`（Bearer API Key 鉴权），
 * 将内部富事件流（content_delta / tool_call / tool_result / status /
 * subagent_progress / error / done）归一化为统一的 dict 事件流。
 * 外部服务，不经 AI SDK（设计文档 ADR-022）。
 */

/** 归一化事件（对应 Python fastclaw_service._normalize_event 输出）。 */
export type FastClawEvent =
  | { type: 'content_delta'; data: { delta: string } }
  | { type: 'content'; data: { delta: string } }
  | { type: 'tool_call'; data: { id: string; name: string; arguments: string } }
  | { type: 'tool_result'; data: { id: string; name: string; result: string } }
  | { type: 'status'; data: { message: string } }
  | { type: 'subagent_progress'; data: Record<string, unknown> }
  | { type: 'error'; data: { message: string } }
  | { type: 'done'; data: {} };

export interface FastClawRuntimeConfig {
  base_url: string;
  api_key: string;
  agent_id: string;
  /** 上游 end-user 标识（FastClaw 据此隔离会话/记忆/用量）。 */
  end_user: string;
}

/** FastClaw 工作区文件（FastClaw agent 相对路径，如 sessions/<chatId>/foo.png）。 */
export interface FastClawWorkspaceFile {
  path: string;
  size: number;
  mtimeMs?: number;
}

export class FastClawAgentError extends Error {
  constructor(message: string, cause?: unknown) {
    super(message, { cause });
    this.name = 'FastClawAgentError';
  }
}

/** 连接阶段超时 15s（读阶段不设上限，由调用方按空闲超时中止）。 */
const CONNECT_TIMEOUT_MS = 15_000;

/** agent 名字解析 TTL：成功 5 分钟 / 失败 30 秒（对应 Python _AGENT_NAME_CACHE_TTL / _AGENT_NAME_FAIL_TTL）。 */
const AGENT_NAME_CACHE_TTL_MS = 5 * 60 * 1000;
const AGENT_NAME_FAIL_TTL_MS = 30 * 1000;
const agentNameCache = new Map<string, { name: string | null; expiresAt: number }>();

/** 从 agent 最终文本中提取第一张图片 URL（支持 markdown 图片语法与裸 URL）。 */
const IMAGE_URL_RE = /!\[[^\]]*\]\((https?:\/\/[^)\s]+)\)|(https?:\/\/[^\s)\]]+\.(?:png|jpe?g|webp|gif)[^\s)\]]*)/;

export function extractImageUrl(text: string): string | null {
  if (!text) return null;
  const m = IMAGE_URL_RE.exec(text);
  if (!m) return null;
  return m[1] ?? m[2] ?? null;
}

export class FastClawAgentService {
  /** 调用 agent 并产出归一化事件流（AsyncGenerator）。
   *  @param signal 调用方中止信号（客户端断连/用户停止时终止底层 SSE 流）。 */
  async *runAgent(
    config: FastClawRuntimeConfig,
    message: string,
    sessionKey: string,
    images?: string[],
    params?: Record<string, unknown>,
    signal?: AbortSignal
  ): AsyncGenerator<FastClawEvent, void, unknown> {
    if (!config.base_url || !config.api_key || !config.agent_id) {
      throw new FastClawAgentError('FastClaw Agent 配置不完整（base_url / api_key / agent_id）');
    }

    const url = config.base_url.replace(/\/+$/, '') + '/api/chat/stream';
    const body: Record<string, unknown> = {
      agentId: config.agent_id,
      sessionId: sessionKey,
      message,
    };
    if (images?.length) body.imageUrls = images;
    if (params) body.params = params;

    const headers: Record<string, string> = {
      Authorization: `Bearer ${config.api_key}`,
      'Content-Type': 'application/json',
      Accept: 'text/event-stream',
    };
    if (config.end_user) headers['X-Fastclaw-End-User'] = config.end_user;

    // 连接阶段超时：手动管理，fetch 完成后立即清除，不影响后续流读取
    const connectAbort = new AbortController();
    const connectTimer = setTimeout(() => connectAbort.abort(), CONNECT_TIMEOUT_MS);
    // 合并连接超时与调用方中止信号（客户端断连/用户停止）
    const fetchSignal = signal
      ? AbortSignal.any([connectAbort.signal, signal])
      : connectAbort.signal;

    let resp: Response;
    try {
      resp = await fetch(url, {
        method: 'POST',
        headers,
        body: JSON.stringify(body),
        signal: fetchSignal,
      });
    } catch (err) {
      throw new FastClawAgentError(`FastClaw Agent 调用失败: ${messageOf(err)}`, err);
    } finally {
      clearTimeout(connectTimer); // 连接完成（成功或失败），清除连接超时
    }
    if (!resp.ok) {
      let detail = '';
      try {
        detail = (await resp.text()).slice(0, 300);
      } catch {
        /* ignore */
      }
      throw new FastClawAgentError(`FastClaw /api/chat/stream 返回 HTTP ${resp.status}: ${detail}`);
    }
    if (!resp.body) throw new FastClawAgentError('FastClaw /api/chat/stream 返回空响应体');

    // SSE 解析：累积 data: 行，空行成块
    const reader = resp.body.getReader();
    const decoder = new TextDecoder();
    let buffer = '';
    let sawDelta = false;
    let sawEvent = false;

    const emit = (payload: string): FastClawEvent[] => {
      const out: FastClawEvent[] = [];
      for (const evt of normalizeEvent(payload)) {
        // FastClaw 先流式发 content_delta，再在本轮结束补发完整 content；
        // 若本轮已收到增量，跳过该条完整 content，避免前端重复拼接文本。
        if (evt.type === 'content_delta') sawDelta = true;
        else if (['tool_call', 'tool_result', 'status', 'subagent_progress'].includes(evt.type)) {
          sawDelta = false;
        }
        if (evt.type === 'content' && sawDelta) continue;
        out.push(evt);
      }
      return out;
    };

    try {
      for (;;) {
        let chunk: { done: boolean; value?: Uint8Array };
        try {
          chunk = await reader.read();
        } catch (err) {
          // 调用方主动中止（客户端断连/用户停止）→ 静默结束，不报错
          if (signal?.aborted) return;
          throw new FastClawAgentError(`FastClaw Agent 流读取失败: ${messageOf(err)}`, err);
        }
        if (chunk.done) break;
        buffer += decoder.decode(chunk.value, { stream: true }).replace(/\r\n/g, '\n');
        const blocks = buffer.split('\n\n');
        buffer = blocks.pop() ?? '';
        for (const block of blocks) {
          const dataLines: string[] = [];
          for (const line of block.split('\n')) {
            if (line.startsWith('data:')) dataLines.push(line.slice(5).replace(/^ /, ''));
          }
          if (dataLines.length === 0) continue;
          const payload = dataLines.join('\n');
          for (const evt of emit(payload)) {
            sawEvent = true;
            yield evt;
          }
        }
      }
    } finally {
      reader.cancel().catch(() => {}); // 确保底层 TCP 连接释放
    }
    // 流自然结束（无 done 事件时补发，保证调用方收尾逻辑统一）
    if (!sawEvent) {
      yield { type: 'done', data: {} };
    }
  }

  /** 列出该 API Key 可访问的 FastClaw agent（admin 页面拉取候选）。 */
  async listAgents(baseUrl: string, apiKey: string, endUser = ''): Promise<unknown[]> {
    const base = baseUrl.replace(/\/+$/, '');
    const headers: Record<string, string> = {
      Authorization: `Bearer ${apiKey}`,
      Accept: 'application/json',
    };
    if (endUser) headers['X-Fastclaw-End-User'] = endUser;
    // 1) dashboard 接口：带真实名字（AgentRecord.name）
    try {
      const resp = await fetch(base + '/api/agents', { headers, signal: AbortSignal.timeout(CONNECT_TIMEOUT_MS) });
      if (resp.ok) {
        const data = (await resp.json()) as { agents?: unknown[] };
        if (Array.isArray(data.agents) && data.agents.length > 0) return data.agents;
      }
    } catch {
      /* 走回退分支 */
    }
    // 2) 上游接口回退：name 与 id 相同
    try {
      const resp = await fetch(base + '/v1/agents', { headers, signal: AbortSignal.timeout(CONNECT_TIMEOUT_MS) });
      if (!resp.ok) throw new FastClawAgentError(`FastClaw /v1/agents 返回 HTTP ${resp.status}`);
      const data = (await resp.json()) as { agents?: unknown[] };
      return Array.isArray(data.agents) ? data.agents : [];
    } catch (err) {
      if (err instanceof FastClawAgentError) throw err;
      throw new FastClawAgentError(`连接 FastClaw 失败: ${messageOf(err)}`, err);
    }
  }

  /** 解析 agent_id 对应的 FastClaw 真实名字（AgentRecord.name，如 "Xulei"）。
   *
   * 带进程内 TTL 缓存（成功 5 分钟 / 失败 30 秒）；FastClaw 不可达或未找到时
   * 返回 null 且短时间内不重试，用于给存量配置回填可读名字（best-effort，不抛错）。
   */
  async resolveAgentName(baseUrl: string, apiKey: string, agentId: string): Promise<string | null> {
    if (!agentId) return null;
    const key = `${baseUrl.replace(/\/+$/, '')}\u0000${apiKey}\u0000${agentId}`;
    const now = Date.now();
    const cached = agentNameCache.get(key);
    if (cached && cached.expiresAt > now) return cached.name;
    let name: string | null = null;
    try {
      const agents = await this.listAgents(baseUrl, apiKey);
      for (const a of agents) {
        if (a && typeof a === 'object' && (a as { id?: unknown }).id === agentId) {
          const candidate = (a as { name?: unknown }).name;
          // 仅接受与 id 不同的可读名字：/v1/agents 回退路径 name==id，
          // 此时视为未解析（返回 null），避免把 agt_xxx 当名字持久化
          if (typeof candidate === 'string' && candidate && candidate !== agentId) {
            name = candidate;
          }
          break;
        }
      }
    } catch {
      name = null;
    }
    agentNameCache.set(key, {
      name,
      expiresAt: now + (name ? AGENT_NAME_CACHE_TTL_MS : AGENT_NAME_FAIL_TTL_MS),
    });
    return name;
  }

  /**
   * 列出当前会话工作区文件（GET /api/agents/{id}/files?sessionId=）。
   * FastClaw 已在服务端按会话/归属做 scoping，这里再做一层「仅当前会话子前缀」过滤兜底，
   * 避免上游布局变更或多余会话混入时把别的会话/agent 根文件透给前端。
   */
  async listSessionFiles(
    config: FastClawRuntimeConfig,
    sessionId: string
  ): Promise<FastClawWorkspaceFile[]> {
    if (!config.base_url || !config.api_key || !config.agent_id || !sessionId) return [];
    const base = config.base_url.replace(/\/+$/, '');
    const url = `${base}/api/agents/${encodeURIComponent(config.agent_id)}/files?sessionId=${encodeURIComponent(sessionId)}`;
    const headers: Record<string, string> = {
      Authorization: `Bearer ${config.api_key}`,
      Accept: 'application/json',
    };
    if (config.end_user) headers['X-Fastclaw-End-User'] = config.end_user;
    let resp: Response;
    try {
      resp = await fetch(url, { headers, signal: AbortSignal.timeout(CONNECT_TIMEOUT_MS) });
    } catch (err) {
      throw new FastClawAgentError(`FastClaw 工作区文件列表失败: ${messageOf(err)}`, err);
    }
    if (!resp.ok) throw new FastClawAgentError(`FastClaw 工作区文件列表返回 HTTP ${resp.status}`);
    let data: { files?: Array<{ path?: unknown; size?: unknown; modTime?: unknown }> };
    try {
      data = (await resp.json()) as typeof data;
    } catch {
      return [];
    }
    const prefix = `sessions/${sessionId}/`;
    const out: FastClawWorkspaceFile[] = [];
    for (const f of Array.isArray(data.files) ? data.files : []) {
      const p = typeof f?.path === 'string' ? f.path : '';
      if (!p.startsWith(prefix)) continue; // 仅当前会话
      const size = typeof f.size === 'number' && Number.isFinite(f.size) ? f.size : 0;
      out.push({
        path: p,
        size,
        mtimeMs: typeof f.modTime === 'number' && Number.isFinite(f.modTime) ? f.modTime * 1000 : undefined,
      });
    }
    return out;
  }

  /**
   * 校验一个 FastClaw 工作区文件路径属于当前会话（防跨会话/目录穿越）。
   * 返回规范化后的路径，非法返回 null。
   */
  static sessionFilePath(sessionId: string, filePath: string): string | null {
    const p = String(filePath ?? '').trim();
    const prefix = `sessions/${sessionId}/`;
    if (!p.startsWith(prefix) || p.includes('..')) return null;
    return p;
  }

  /**
   * 拉取 FastClaw 工作区文件字节（GET /api/agents/{id}/files/{path...}）。
   * 由调用方把 resp.body 流式转发给浏览器。路径须先经 sessionFilePath 校验，
   * 否则返回 null（404）。
   */
  async fetchSessionFile(
    config: FastClawRuntimeConfig,
    sessionId: string,
    filePath: string
  ): Promise<Response | null> {
    const p = FastClawAgentService.sessionFilePath(sessionId, filePath);
    if (!p || !config.base_url || !config.api_key || !config.agent_id || !sessionId) return null;
    const base = config.base_url.replace(/\/+$/, '');
    // 逐段编码：路径含空格/中文时仍被 FastClaw {path...} 路由正确解析，且保持层级 '/'。
    const encPath = p.split('/').map((seg) => encodeURIComponent(seg)).join('/');
    const url = `${base}/api/agents/${encodeURIComponent(config.agent_id)}/files/${encPath}`;
    const headers: Record<string, string> = {
      Authorization: `Bearer ${config.api_key}`,
      Accept: '*/*',
    };
    if (config.end_user) headers['X-Fastclaw-End-User'] = config.end_user;
    let resp: Response;
    try {
      resp = await fetch(url, { headers, signal: AbortSignal.timeout(CONNECT_TIMEOUT_MS) });
    } catch (err) {
      throw new FastClawAgentError(`FastClaw 工作区文件下载失败: ${messageOf(err)}`, err);
    }
    if (!resp.ok) return null;
    return resp;
  }
}

/** 解析单个 SSE payload 为归一化事件（对应 Python _normalize_event）。 */
function normalizeEvent(payload: string): FastClawEvent[] {
  let evt: unknown;
  try {
    evt = JSON.parse(payload);
  } catch {
    return []; // 非 JSON 行（杂讯）忽略，不中断流
  }
  if (typeof evt !== 'object' || evt === null) return [];
  const e = evt as { type?: string; data?: unknown };
  const data = (typeof e.data === 'object' && e.data !== null ? e.data : {}) as Record<string, any>;
  switch (e.type) {
    case 'content_delta': {
      const delta = typeof data.delta === 'string' ? data.delta : '';
      return delta ? [{ type: 'content_delta', data: { delta } }] : [];
    }
    case 'content':
    case 'message': {
      const content = typeof data.content === 'string' ? data.content : typeof data.text === 'string' ? data.text : '';
      return content ? [{ type: 'content', data: { delta: content } }] : [];
    }
    case 'tool_call':
      return [{ type: 'tool_call', data: { id: str(data.id), name: str(data.name), arguments: str(data.arguments) } }];
    case 'tool_result':
      return [{ type: 'tool_result', data: { id: str(data.id), name: str(data.name), result: str(data.result) } }];
    case 'tool_progress':
    case 'status':
      return [{ type: 'status', data: { message: typeof data.message === 'string' ? data.message : typeof data.tool === 'string' ? data.tool : JSON.stringify(data) } }];
    case 'subagent_progress':
      return [{ type: 'subagent_progress', data }];
    case 'error':
      return [{ type: 'error', data: { message: typeof data.message === 'string' ? data.message : JSON.stringify(data) } }];
    case 'done':
      return [{ type: 'done', data: {} }];
    default:
      return [];
  }
}

function str(v: unknown): string {
  return typeof v === 'string' ? v : v == null ? '' : String(v);
}

function messageOf(err: unknown): string {
  return err instanceof Error ? err.message : String(err);
}

/** 单例（对应 Python 模块级 fastclaw_agent_service）。 */
export const fastclawAgentService = new FastClawAgentService();
